server/EXCEPTIONS.md

```diff
@@ -0,0 +1,168 @@
+
+Only the HttpException caught and thrown closest to or on the controller
+endpoint will be returned for a given request. To prevent confusion over which
+backend layer should provide the most detailed error handling, this document prescribes generating new HTTPExceptions in services, and simply relaying them in controllers.
+
+## General Errors / Preventing Gaps
+Since we are trying to `throw new HttpException` at the service level, we may
+want to write a generic, controller-level catch-all for errors. This will help catch
+any uncaught errors from our services. Services are where we house the bulk
+of our business logic for processing requests, so they have a large surface
+area over which an error may not yet be covered by (occur outside of) a
+try-catch.
+
+```js
+// controllers/my-project.controller.ts
+@Get('/')
+async getProject(@Session() session) {
+  try {
+    return this.myProjectService.getProject();
+  } catch (e) {
+    if (e instanceof HttpException) {
+      throw e;
+    else {
+      // A catch-all, providing a general error
+      throw new HttpException({
+        code: 'GET_PROJECT_UNKNOWN',
+        message: 'An unknown server error occured while getting projects',
+      }, HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
+}
+```
+
+## HTTP Exception Filter
+
+Our nest backend has a custom [Nest exception filter](https://docs.nestjs.com/exception-filters#throwing-standard-exceptions) in place to catch all thrown HttpExceptions and format them before relaying them to the frontend. It will wrap the errors inside an `errors` array in an object:
+
+```json
+  {
+    errors: [ ...thrownHttpExceptions]
+  }
+```
+
+This means in our code we don't need to manually wrap the error. Below is what manually wrapping the error would look like, but our HttpException does this for us:
+
+```js
+// controllers/my-project.controller.ts
+@Get('/')
+async getProject(@Session() session) {
+  try {
+    ...
+  } catch (e) {
+    if (e instanceof HttpException) {
+      res.status(401).send({ errors: [e] });
+    } else {
+      console.log(e);
+
+      res.status(500).send({ errors: [e] });
+    };
+}
+```
+
+The custom HTTP Exception filter helps format the error response so that the frontend
+can easily ingest and display the errors.
+
+## Talking to the frontend
+
+Our frontend has an adapter that automatically picks up and
+attaches errors from the backend to frontend models. We have also
+written error pages to automatically display the errors, so long
+as the incoming errors conform to the format below -- achieved by
+following the example for generating errors in the Quickstart section. 
+
+```json
+{
+  errors: [
+    response: {
+      message: "our error message", 
+      code: OUR_ERROR_CODE, // We also determine this value
+    },
+    status: [401|500|etc...],
+    message: "our error message"
+  ]
```

Comment:
Could the example strings in this code sample help clarify the difference in the two `message` values?

Reply:
They're actually the same! The value of response.message is copied over to the top-level message property